datalad/interface/rerun.py

```diff
@@ -0,0 +1,277 @@
+            constraints=EnsureDataset() | EnsureNone()),
+        message=save_message_opt,
+        branch=Parameter(
+            metavar="NAME",
+            args=("-b", "--branch",),
+            doc="create and checkout this branch before rerunning the commands.",
+            constraints=EnsureStr() | EnsureNone()),
+        onto=Parameter(
+            metavar="base",
+            args=("--onto",),
+            doc="""start point for rerunning the commands.  If not specified, commands
+            are executed at HEAD.  This option can be used to specify
+            an alternative start point, which will be checked out with
+            the branch name specified with --branch or in a detached
+            state otherwise.  As a special case, an empty value for
+            this option means to use the parent of the first commit
+            (e.g., 'a' for the range 'a..b' or 'a^' for the revision
+            'a') as the starting point.""",
+            constraints=EnsureStr() | EnsureNone()),
+        root=Parameter(
+            args=("--root",),
+            action="store_true",
+            doc="""rerun commands from all commits reachable from revision rather than
+            running only the command from the revision.  This flag is
+            incompatible with using a range for the revision the
+            argument. In other words, run all the commands that would
+            be shown by `git log <revision>`."""),
+        # TODO
+        # --list-commands
+        #   go through the history and report any recorded command. this info
+        #   could be used to unlock the associated output files for a rerun
+    )
+
+    @staticmethod
+    @datasetmethod(name='rerun')
+    @eval_results
+    def __call__(
+            revision="HEAD",
+            dataset=None,
+            branch=None,
+            message=None,
+            onto=None,
+            root=False):
+
+        ds = require_dataset(
+            dataset, check_installed=True,
+            purpose='rerunning a command')
+
+        lgr.debug('rerunning command output underneath %s', ds)
+
+        from datalad.tests.utils import ok_clean_git
+        try:
+            ok_clean_git(ds.path)
+        except AssertionError:
+            yield get_status_dict(
+                'run',
+                ds=ds,
+                status='impossible',
+                message=('unsaved modifications present, '
+                         'cannot detect changes by command'))
+            return
+
+        err_info = get_status_dict('run', ds=ds)
+        if not ds.repo.get_hexsha():
+            yield dict(
+                err_info, status='impossible',
+                message='cannot re-run command, nothing recorded')
+            return
+
+        if branch and branch in ds.repo.get_branches():
+            yield get_status_dict(
+                "run", ds=ds, status="error",
+                message="branch '{}' already exists".format(branch))
+            return
```

Comment:
Why is this a problem?

Reply:
because 
```
  -b NAME, --branch NAME
                        create and checkout this branch before rerunning the
                        commands. Constraints: value must be a string
```
i.e. that will be the branch from the starting point where to record changes if any happen.  If it already exists, and you want to "replay on top", use `--onto branch` ?

Reply:
> If it already exists, and you want to "replay on top", use --onto branch ?

Yeah, although that will do things in a detached state rather than
extending the branch :/

Perhaps it'd be better to let the user specify detached/not based on
how they give the argument?  `--onto=branch` would check out the
branch, whereas `--onto=branch^0` would check out in the detached
state (and of course a `--onto=commit` would always be detached).  But
that's relying on a Git syntax that a lot of users probably won't be
familiar with (or making them use hashes) if they want a detached
state.

Reply:
> Why is this a problem?

I made it work this way because I think it makes the interaction with
`--onto` simpler.  If we handled existing branches here, we'd need to
consider resetting the branch to honor the `--onto` argument.